Improvements to relational database handling in Orleans

veikkoeeva opened the pull request:
Starts a fix for #255.

An attempt to improve relational database handling in Orleans. If this approach is all right, I'd appreciate help either someone pointing a place to move the files or pulling this PR in and moving the files around. After which the work would continue with the relocated files (possibly in a separate PR).

#### Points of note:

1. Code is not used anywhere at the moment. A feasible approach to make use of it initially could be use in tests by modifying [Local.testsetting](https://github.com/dotnet/orleans/blob/master/src/Local.testsettings#L6) by adding [CreateTables.sql](https://github.com/dotnet/orleans/blob/master/src/OrleansProviders/SQLServer/CreateTables.sql) to the set of included items in tests [MembershipTablePluginTests](https://github.com/dotnet/orleans/blob/master/src/TesterInternal/StorageTests/MembershipTablePluginTests.cs#L38) and create database as appropriate for testing (code follows) and add the tables in the script in (one has to remove the ``DELETE`` clauses, likely also the ``SELECT`` clauses.
2. Laying this foundation should support all databases with ADO.NET providers.
3. Should make it easier to add new relational backends so that they are easily testable. This is achieved chiefly with special template queries. See ``RelationalConstants``.
4. Should not hinder adding further features such as using stored procedures or reading
    multiple resultsets.
5. Is asynchronous all the way.
6. Does not have logging facilities added.
7. Includes a few reflection utilities to make use easier, though hand-tune queries are possible too and likely should be favoured. An example of such is at ``ExistsAsync``.
8. Current Orleans queries aren't added to the newer code base. One approach would be to turn them to extensions methods.
9. Related to point number eight, currently data would be inserted in a loop, but a more efficient would be multi-row ``INSERT INTO VALUES``, which is possible, but having a few methods to assembly the query string from a provided collection string would make it nicer. On the TODO list.

#### Additional notes:
1. Perhaps it would be useful to create a new project, e.g. ``Orleans.DataStorage`` wherein to put data storage related functionality.
2. Perhaps it would be useful to define an interface according to HTTP verbs (GET/PUT/PATCH/DELETE). This way adding any storage backend would look like being easier. Caveat emptor, I haven't studied this in depth.
3. Perhaps it would be useful to create a special ``ReflectionCache`` component, which would include the source of type and a timestamp when added.
4. Perhaps it would be feasible to add a project like ``Orleans.UnitTests``, which would contain clearly unit test like functionality.

In the following debugger and/or open database might be handy. This is a simpble usage example ("works on my machine")

```C#
     using System.Data.Common;
     using System.Linq;
     using UnitTests.StorageTests.Utilities;


    namespace OrleansTest
    {
    public struct Test
    {
        public int configuration_id { get; set; }

        public string description { get; set; }
    }

        
    class Program
    {
        static void Main(string[] args)
        {
            var sqlServer = RelationalConstants.GetRelationalConstants().First(i => i.InvariantName == WellKnownRelationalInvariants.SqlServer);
            var db = GenericRelationalStorage.CreateInstance(sqlServer.InvariantName, sqlServer.DefaultConnectionString);

            //Read some data from the master database.
            var r1 = db.ReadAsync<Test>("SELECT * FROM sys.configurations WHERE configuration_id < @Id", new { Id = 200 }).Result;

            //Create a new database if it doesn't exist already. This needs to be done with a valid connection string.
            //The one in default template connects to master database, which always exists.
            if (!db.ExistsAsync("Testtest").Result)
            {
                db.CreateAsync("Testtest").Wait();
            }

            //Now that there's a new database, let's connect to it...
            var csb = new DbConnectionStringBuilder();
            csb.ConnectionString = sqlServer.DefaultConnectionString;
            csb["Initial Catalog"] = "Testtest";
            db = GenericRelationalStorage.CreateInstance(sqlServer.InvariantName, csb.ConnectionString);

            //Create a new table. Note to refactor .ExecuteAsync so that adding "null" isn't needed. INFORMATION_SCHEMA should be a somewhat implemented standard metadata table...
            var r2 = db.ExecuteAsync("IF EXISTS(SELECT * FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = 'Test') DROP TABLE Test;", null).Result;
            var r3 = db.ExecuteAsync("CREATE TABLE Test(Id INT PRIMARY KEY IDENTITY(1, 1) NOT NULL, Data NVARCHAR(256) NOT NULL);", null).Result;            
            var r4 = db.ExecuteAsync("INSERT INTO Test(Data) VALUES(@Data);", new { Data = "Some data 1..." }).Result;
            var r5 = db.ExecuteAsync("INSERT INTO dbo.Test(Data) VALUES(@Data);", new { Data = "Some data 2..." }).Result;
            var r6 = db.ExecuteAsync("DELETE FROM Test WHERE Id = @Id;", new { Id = 1 }).Result;
        }
        }
    }
```

veikkoeeva commented:
@jthelin Towards the end of the week, I think I'll refactor a bit (see the point about null parameter). I may also try to switch the test code to use the new bits and when that works and assuming this approach is OK, we can move to change the Orleans itself to use this new code. I may also add the current queries as extensions methods or by some other mechanism (it would look like ``CreateAsync`` there for the caller, with appropriate parameters, of course).

As a note, the example code I put there assumes a DB process is running, which is with a normal VS installation (e.g. LocalDb). I'm not sure about Orleans CI server setup. I might provide a separate PR for a code that cheks if appropriate processes are running and message with them. It could check that Storage Emulator is running before tests too.

There's also an issue getting the connector libraries, as ADO.NET facilities try to search either from a GAC or from a pointer given in a ``.config file``, one of which is not the Orleans configuration files. GAC isn't such a problem, but there may be a need to use reflection or otherwise point the probing to appropriate places (and one issue would the schema of the Orleans config file, I think, I haven't investigated).

I may also install MySQL to test that too. I got a bit of interrim headache, so I wasn't able to that. :)

jthelin commented:
Hi @veikkoeeva - Just to clarify, are you looking to merge this first tranche into master at this stage or keep iterating? 

Everything looks good to me so far [modulo a few minor code review feedback items which I posted above] I think I can see where it is all converging on, but it is a little hard to get the full picture until the first end-to-end test case is added that uses this new test framework. ;)

As Sergey mentioned, we have some internal test cases we used in the past for testing some low level storage / membership provider functionality but the challenge has always been getting them to run reliably using SQL Server libraries on the unattended build servers which don't have Visual Studio installed. 

Seems like this new db test infrastructure can self-bootstrap through NuGet package references and a little metadata using your new test framework code, which would definitely be a great step forward! :)

veikkoeeva commented:
@jthelin Let's see, I was rather tired when I added this... :) I think I would like to know where to put the files. They don't belong to the test project if the intention is to use them from the actual grain saving code too.

One path forward would be that I leave the files as is there and refactor the tests to use this new code. If that goes well, we'll think a place for them in some other Orleans project. It wouldn't hurt to test this code itself too, I have a full intention to do that. I think what I try to test is that I create a database, check creation succeeds, load in the ``.sql´´ script, see it succeeds, load in some data to the tables and look the right data is read back also. After that there would be the other tests.

Yes well, this code assumes there's a DB process running. Otherwise yes, the tests should be runnable straight by cloning the repo and updating Nuget packages. In Gitter there was little pondering already if there should be code that checks if the appropriate DB processes are running. In any event, MySQL is easy to install on Windows, likewise LocalDb or SQL Server Express too, which would give two different databases to test. It shouldn't be hard to add yet more databases to the matrix, say, Oracle, if one wants to (perhaps privately).

Anyway, I'll be off to sleep soon, but take this as :+1: for the direction and go towards making the tests run (at least locally, the CI machine needs a running DB server) on the new code.

veikkoeeva commented:
@jthelin I updated the relational tests to run on this new system. Would you or someone else like to take a look where to refactor the files that are currently in ``TestInternal\Utilities``? As far as I'm concerned, if this looks good, we could put the code in, refactor the new classes to correct projects and continue with a new PR on moving the actual Orleans code to use this code (e.g. ``SqlMembershipTable``). What do you think? The only risk I see here is that I altered ``CreateTables.sql`` and ``Local.testsettings``.

Caveat emptor here is that I haven't still run this code on other DBs. I tried a quick MySQL install (and I'd need to add some constant MySQL SQL queries too to the code), but the new and improved Oracle setup experience sucked enough that I think I'll just go to bed. :)

jthelin commented:
A quick update from my side: In general these changes look good, and I think the best approach is to the current crawl-walk-run approach of piloting these changes in test cases first and then eventually we can move the supporting DB utility code into the main runtime.

I have been trying running them as part of our internal test suites, and there is some glitches causing some Azure test cases to fail which may or may not be related to this change but I am still investigating. .

The 3 MT_*_SqlServer test cases all run fine on my development machine, and would be good candidates to add to BVT test suite because the 3 tests run in ~9 seconds total. At the moment I cannot confirm that they run ok on the build servers until I have fixed the Azure test problems above.

Stay tuned ... :)

veikkoeeva commented:
@jthelin I gittered (is that a word?), but I've been too busy to follow through with it. :)

What you propose seems like the best course of action. I did a short demo for myself last evening (about eight hours ago) where I added multiple resultsets, transactions and thought about ``INSERT INTO VALUES`` (multiple rows at once) and fixed some false information on the interface comments. These are mostly API design issues currently.

I think I'll make some progress towards adding them to this PR in an evening or two. My plan after these changes was/is to port/copy the current queries from the main run-time to the test project. That is, to make the current queries run on test project.

I did also some research on performance and stability.``INSERT INTO VALUES`` (multiple rows at once) in addition having different syntax on Oracle compared to others has some performance characteristics worth being aware about. I found articles with measurements that using parameters is less performant than using non-parameterized string in that case. I'm not entirely sure if there there's switch in some DB where a one can [force only parameterized](https://technet.microsoft.com/en-us/library/ms175037%28v=sql.105%29.aspx) queries to be accepted (probably not).

In any event, I think I'm not doing an ORM with all the bells and whistles here and won't be adding everything I can think of, but concentrate on some things I believe are needed to give room to work with multiple DBs in near optimal way and maintain the possibility for hand-crafted queries.

For someone reading here, the reason I'd like to have a reasonable abstraction over the relational storage surface before plunging to the table design is that the DBs have different characteristics and some minimal shiedling abstractions looks like the way to go to support DB features without a least-common denominator (yeah, licences cost money, bang for the buck :)) way and then make the system testable and sufficiently robust from Orleans perspective without edge cases (which would show as troubleshooting questions).

veikkoeeva commented:
@jthelin Some more code added. Most notably

* Multiple resultsets 
* Multi-record insert
* Transactions
* Tweaks to database creation

When tweaking ``CreateTables.sql`` I fully assumed it is SQL Server only. I think we need to rename it at some point to something like ``SqlServer_CreateTables.sql``.

I think the following things need to be done next:

1. Introduce views. This way the current queries would work by just replacing table names with views names and then hiding the joins in some queries behind these new views. This may need some other changes in the Orleans database related code too (I'm not sure how the lists are build etc.). The original issue has more ideas why views are for the win.
2. I was thinking there could be new helper class to create Orleans specific queries. Doing it by reflection likely works, but is naturally slower than would be by handcrafting them. I think ``SELECT`` ones should be parameterized, multi-record ``INSERT`` ones not so (though possible). As for a note, exampe is [here](https://github.com/veikkoeeva/orleans/commit/2b3419d081f9a2930fcf1053d97e709ad74130ff#diff-7c109863ac54d6609e966d8c07b3bfa2R276) where it's like an ordinary result set reading. Notable thing is that in case there are multiple result sets, a count of the result set is transmitted.
3. If at all possible, more tests to relation storage.
4. There's always something more.

@jthelin Let me know if you are working on something related to this or if you find something about the glitches. Otherwise, the usual caveats I may make progress slowly, so anyone chiming in is welcome to do so. :)

Oh, almost forgot. There are examples in the test code, but this works like follows (with reflection):

```C#
public struct Test
{
    public int configuration_id { get; set; }
    public string description { get; set; }
}

public struct Test2
{
    public string Data { get; set; }
    public DateTime? D1 { get; set; }
    public DateTimeOffset? Do { get; set; }
}

using(var t = db.CreateTransactionScope())
{
    var r0 = db.ReadAsync<Test>("SELECT * FROM sys.configurations").Result;
    var r1 = db.ReadAsync<Test>("SELECT * FROM sys.configurations WHERE configuration_id < @Id", new { Id = 200 }).Result;

   t.Complete();
}

//A table is needed, could (should) be run inside a transaction too...
var cr = db.ExecuteAsync("IF NOT EXISTS(SELECT * FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = 'Test') " +
                "CREATE TABLE Test(Id INT PRIMARY KEY IDENTITY(1, 1) NOT NULL, Data NVARCHAR(256), DefaultDateTime DATETIME2(7) DEFAULT SYSUTCDATETIME(), D1 DATETIME2(7), Do DATETIMEOFFSET);").Result;

using(var t = db.CreateTransactionScope())
{
    var params2 = new[] { new Test2 { Data = "X", D1 = null, Do = DateTimeOffset.Now }, new Test2 { Data = "Y", D1 = DateTime.Now.ToLocalTime(), Do = DateTimeOffset.UtcNow } };
    var q = db.CreateMultipelInserIntoQuery("Test", params2);
    var rm0 = db.ExecuteAsync(q).Result;
    //Multiple result sets...
    var r0 = db.ReadAsync<Test2>("SELECT * FROM Test;SELECT * FROM Test;").Result;
}
```

veikkoeeva commented:
@jthelin I added functionality to query, update and insert table membership data (not all of it, but it's more on the hand-tuned side). It's somewhat difficult to test (I really didn't for the time being) without starting to replace [MembershipTable](https://github.com/dotnet/orleans/blob/master/src/Orleans/Messaging/SqlMembershipTable.cs) members. What do you think, should we (you? :)) move the code off of the testing utilities directory (appropriate namespace refactoring etc.) somewhere (where?) or what should be the plan? I should note the old code should continue working as is until redone. Yes, still not too late to rethink this approach I'm taking. :-P

I have an idea I could use the new functions in the current three tests so that I check I can read the data that's produced by the tests when they call the existing functionality.

You start to see there a few things:
* I introduce a view to query membership state. I expect these view "interfaces" to be very stable from database to database (that is, equivalent data types and columns, queries themselves may vary).
* There's functionality to insert data to database. This is somewhat database specific, in fact a bit version specific (e.g. how to do upsert pre SQL Server 2008), but it can all be handled if necessary. The code can query database version (indeed, even database vendor) and select queries appropriately (see ``OrleansDatabaseInfo``). Currently the queries are defined in the code, but I was thinking to store them also to database (currently I'm inclined to not to store in the database).
* I changed the membership ``LEFT JOIN`` to ``INNER JOIN`` as it look like making sense as it were. If that is indeed wrong, a similar functionality to what the code is doing with multiple queries can be achieved by a ``CASE`` in ``SELECT`` that checks if the version ID is null and then retrieving it separately. I think I'll redo some of that when I get to refactoring more of the database itself. How the tables are and so forth doesn't feel all right as-is currently.

I *may* busy be enough that I might not be able to move the code measurably forwards during the next week, which likely means the weekend after next weekend. I'm not sure yet. That's one of the reasons I'm asking how should we proceed from here, as it would start to make sense to start moving the code to Orleans from the testing, gradually replace current functionality and do better testing. In fact I was thinking to add queuing related stuff too in some not-so-distant future. We'll see. Smaller worries first, like that I wasn't able to get statistics to the database...

jthelin commented:
Thanks @veikkoeeva -- you are really getting things pulled together here.

I'll take a look, but likely to take a few days due to so many other things on the TODO list currently :crying_cat_face:

veikkoeeva commented:
@jthelin Ah, just going to bed... Sure. Shoot a message if you have something ask. Any guestion is all right.

veikkoeeva commented:
@jthelin A couple of notes that I have sober head. To be specific, I think now it's good time to find a place these in the code unless you or someone else finds something that requires further action. In order not to burden you with too much work and scratching head, here's a short list of reasoning and todo:

* It makes sense to inject a logger to ``GenericRelationalStorage``.
* It may make sense to add ``CancellationToken`` to database operations.
* It may make sense to add an interceptor to the pipeline.

Regarding the interceptor. The idea here is that ``GenericRelationalStorage`` is a thin wrapper around ``System.Data.*`` factories masking away the various factory methods that make the platform libraries more or less vendor agnostic, so interceptor may not be needed because one would in any event write the query itself and call exactly the same methods. Then the ``*Constants`` classes are "query bags" from which a query is chosen depending on task at hand. I took an approach where I created a set of convenience methods that grab a desired query and makes use of ``GenericRelationalStorage``. These can be filled from a database too, if that would make sense. When the database is created, that's a hard requirement to know the gritty details of vendor and a version, so it may make sense to store the used queries there to a special table. There are some limitations to this approach, so initially, I think, it makes sense to have a predefined set of queries defined in code too.

The rest are a lot of convenience functions I used myself and thought it wouldn't hurt to include them there. I have not yet tested with other databases than SQL Server (and that too in a limited capacity), but I think I'll try at least MySQL myself in the testing project. I can't do it currently as Orleans use SQL Server specific functionality (``SqlCommand``, ``SqlConnection`` etc.).

Related to testing, there's a function [GetRelationConstants](https://github.com/dotnet/orleans/pull/347/files#diff-9643281f3d2d42e7399ca5d276deff0cR204) that can be used to enumerate all ADO.NET providers found (either defined in ´´.config`` or in ``machine.config``). The function looks for a vendor specific invariant name and matches database management queries to that. Then, as you can see in the testing project, there are conveniene functions that are being used. In essence one could make sure the database processes are running and run the test set for all of the databases in a loop. I think that is the way forward even with table storage tests as long as the tests aren't table storage specific. That way test added would be useable across all of the storage backends, including higher-level fault handling and circuit breaker logic.

veikkoeeva commented:
@jthelin Ping. If I have time during the next nine days, would it be all right I move the code to some suitable looking place and start replacing the existing ``SqlCommand`` code? This way I could run the the three tests with the new code and fix things that come up.

Just to make sure still, don't hesitate to bring up if there's something you'd like to ask.

ReubenBond commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -67,448 +72,71 @@ public SqlMembershipTable(ClientConfiguration config)
+        async Task<bool> IMembershipTable.UpdateRow(MembershipEntry entry, string etag, TableVersion tableVersion)
         {
-            return Guid.NewGuid().ToString();
+            return await database.UpsertMembershipRowAsync(deploymentId, etag, tableVersion, entry);
```
This function should not be async, just return the task

ReubenBond commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -67,448 +72,71 @@ public SqlMembershipTable(ClientConfiguration config)
-                return new MembershipTableData(memEntries, await GetTableVersion(conn));
-            else
-                return new MembershipTableData(memEntries, new TableVersion(version, versionETag));
+            await database.UpdateIAmAliveTimeAsync(deploymentId, entry);
```
As above: return task instead of awaiting

jthelin commented:
@veikkoeeva - are you still wanting to do more work on this, or ready to merge to master?

veikkoeeva commented:
@jthelin There will be more work. I just got the ``SqlMembershipTable.cs``, along with membership protocol, done. I will convert the other queries too, tidy up places and very likely write some more tests. You should be in Gitter more. :wink: 

In any event, maybe sometime next week.

jthelin commented:
No worries @veikkoeeva . 
Just @mention ping me when you are ready to merge. 
It all LGTM so far :)

Sorry, I keep getting pulled off onto other projects using Twitter data streams, 
so I can't seem to keep up with the Glitter 'firehose' too ;)

veikkoeeva commented:
@jthelin Yeah, this was a bit more work than I anticpated. Just a note that I have a newer version of the code and there's (still) a small bug in querying membership data (NULLs should be checked in the query and not to apply conditions if they are).

I also plugged the SqlServerMembershipTable to the same tests that are used for Azure table, but more details to follow. I'd suspect @shayhatsor would need to do something similar in https://github.com/dotnet/orleans/pull/493. I have an itch to refactor the tests a bit to make plugging new membership providers more easily and perhaps some of the tests should be named with something more like *membership protocol tests*. In any event, just a heads up that I'm working on this and not to be too confident on the existing code (I'll push either today or tomorrow when I'm at home again).

jasonwilliams200OK commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,148 @@ MIT License
 using System.Threading.Tasks;
 
-using Orleans.Messaging;
-using Orleans.Runtime.Configuration;
```
<img src="http://i.stack.imgur.com/71HGg.png" alt="sort usings with system first" />

Please check this option and resort usings of each code file.

jasonwilliams200OK commented on src/Local.testsettings:
```diff
@@ -2,7 +2,7 @@
   <Description>These are default test settings for a local test run.</Description>
   <Deployment>
-    <DeploymentItem filename="Tester\bin\Debug\TestInput\" />
+    <DeploymentItem filename="Tester\bin\Debug\TestInput\" />	
```
:bulb: whitespace at the end of this line.

jasonwilliams200OK commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,148 @@ MIT License
+        private IRelationalStorage database;
+
 
         public async Task InitializeMembershipTable(GlobalConfiguration config, bool tryInitTableVersion, TraceLogger traceLogger)
```
extra linefeed: one is enough :wink:

jasonwilliams200OK commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,148 @@ MIT License
             }
         }
 
+
```
linefeed

jasonwilliams200OK commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,148 @@ MIT License
             return TaskDone.Done;
         }
 
+
```
linefeed

jasonwilliams200OK commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,148 @@ MIT License
         }
 
+
         public bool IsUpdatable
```
linefeed

jasonwilliams200OK commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,148 @@ MIT License
             get { return true; }
         }
 
+
```
linefeed

jasonwilliams200OK commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,148 @@ MIT License
-            "WHERE DeploymentId = @deploymentid";
     }
-}
+}
```
On Windows when using git and the line-ending used are CRLF, please add an extra linebreak to avoid this:

![wrong-eof](https://cloud.githubusercontent.com/assets/2031765/8147734/a9765ae0-1283-11e5-8864-2690ea9b7990.PNG)

jasonwilliams200OK commented on src/Orleans/Storage/DbExtensions.cs:
```diff
@@ -0,0 +1,144 @@
+using System.Data;
+
+
+namespace Orleans.Runtime.Storage.Relational
```
extra linefeed

jasonwilliams200OK commented on src/Orleans/Storage/DbExtensions.cs:
```diff
@@ -0,0 +1,144 @@
+        }
+
+
+        /// <summary>
```
linefeed

veikkoeeva commented:
@jthelin Just a note that the statistics table is yet to be converted as I was looking for a way to test it properly. Is it certain it works currently? I haven't been able to configure it so that I get logs to SQL database tables. The conversion itself shouldn't take that long (an evening or two), but being somewhat certain everything works as supposed might take longer. Same with reminder service, there isn't tests currently, maybe I should add some.

A side note, I undo the changes to `` src/TesterInternal/StorageTests/AzureMembershipTableTests.cs `` before the merge.

veikkoeeva commented:
@jasonwilliams200OK ACK. I'll look into these when this is about to merge. Shouldn't take long anymore, but you never know.

veikkoeeva commented:
Intermediary update: Tests pass, I'm able to start and restart deployment with one or multiple silos and reminders seem to be working (I can add and delete them in which case they show up in DB or are removed). Statistics to go, and I got the configs to work.

gabikliot commented:
Great progress!

sergeybykov commented:
@veikkoeeva I've been trying to migrate our internal liveness tests (including their SQL versions) to Github and ran into some obscure issues with that. Need a bit more time.

veikkoeeva commented:
@jthelin I'm doing some cleanup, adding comments and waiting for #557 tests. I'll try to annotate some parts of the code during the weekend too.

veikkoeeva commented:
@sergeybykov 

> I've been trying to migrate our internal liveness tests (including their SQL versions) to Github and ran into some obscure issues with that. Need a bit more time. 

I didn't notice this (forgot to refresh). What kind of weirdness? I think I need to adjust the tests somewhat so that they are automatically run on LocalDb, so if they are database related, maybe a good course of action is to commit them to master, I pick them up and try to see too. Maybe @gabikliot referred to this already in #557.

gabikliot commented:
Exactly. #557 are additional tests. There is already bunch of membership table related tests for SQL in master. #557 adds more end to end liveness tests. In addition we will improve more of the #557 tests that their Test attribute is currently commented out and that will give even more tests.

veikkoeeva commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,229 @@ MIT License
+            logger.Verbose3("SqlMembershipTable.InitializeMembershipTable called.");
+                                                
+            //TODO: Orleans does not yet provide the type of database used (to, e.g., to load dlls), so SQL Server is assumed.
+            database = RelationalStorageUtilities.CreateGenericStorageInstance(WellKnownRelationalInvariants.SqlServer, config.DataConnectionString);
```
I'm thinking of opening an issue to word out the work needed in more detail. The very least that needs to be done is to transmit ``provider invariantname``, such as *System.Data.SqlClient* to the system. It could be a new attribute in Orleans configuration, but a little additional twist in the matter is that "normally" some of these are controller also in ``.config`` files in the ``system.data`` block. [MSDN Documentation](https://msdn.microsoft.com/en-us/library/dd0w4a2z%28v=vs.110%29.aspx), a fairly well explained [blog post](https://lukewickstead.wordpress.com/2014/02/26/ado-net-connections-and-data-providers/). I think this case warrants to think a solution that should work in any case.

veikkoeeva commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,229 @@ MIT License
-                    }
-                }
+                //TODO: Refactor this to async.
+                return database.ActiveGatewaysAsync(deploymentId).Result;
```
I'm not sure if there's an issue for this already, but this sticks into one's eye a bit. :) This looks like requiring refactoring some common functionality too.

veikkoeeva commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,229 @@ MIT License
-                command.Parameters.Add(new SqlParameter { ParameterName = "@generation", DbType = DbType.Int32, Value = key.Generation });
-
-                return await ProcessResults(conn, command);
+                logger.Verbose("SqlMembershipTable.Gateways failed {0}", ex);
```
I don't know if ``.Verbose`` is the correct level to log membership errors. The tasks are awaited in lower SQL level and the tasks are basically pass-through here, so not all problems are catched here, I think.

veikkoeeva commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,229 @@ MIT License
-            // NOTE: This method is only called from client, so blocking operations should be ok.
-
-            using (var conn = new SqlConnection(connectionString))
+            logger.Verbose3("SqlMembershipTable.GetGateways called.");
```
I added an explicit print of calling membership methods. It doesn't feel like there's harm done (performancewise or otherwise), especially on level three of verbosity.

veikkoeeva commented on src/Orleans/Messaging/SqlMembershipTable.cs:
```diff
@@ -21,489 +21,229 @@ MIT License
+            //not to go through all the way to database to fail a conditional check on etag (which does
+            //exist for the sake of robustness) as mandated by Orleans membership protocol.
+            //Likewise, no update can be done without membership entry.
+            if(tableVersion == null || entry == null)
```
These are sanity checks. No need to go over the wire all the way to the database to get a failure if there's a fatal problem. This shouldn't happen in normal, non-development scenarios, I suppose.

veikkoeeva commented on src/Orleans/Storage/DbExtensions.cs:
```diff
@@ -0,0 +1,192 @@
+        /// <summary>
+        /// An explicit map of type CLR viz database type conversions.
+        /// </summary>
+        static readonly ReadOnlyDictionary<Type, DbType> typeMap = new ReadOnlyDictionary<Type, DbType>(new Dictionary<Type, DbType>
```
I added this to suppor reading database by reflection. Reflection with database isn't used by Orleans itself, but functionality like this makes this Orleans specific implementation a mini-ORM with nearly no added trouble.

veikkoeeva commented on src/Orleans/Storage/DbExtensions.cs:
```diff
@@ -0,0 +1,192 @@
+using System.Data;
+
+
+namespace Orleans.Runtime.Storage.Relational
```
This is currently public and shows to application code. It means someone might a dependency on it. Maybe some other namespace name would be in order? If there's a plan to separate storage code to their respective assemblies, maybe this could reflect that organization already now even if the refactoring would come at later stage.

veikkoeeva commented on src/Orleans/Storage/DbExtensions.cs:
```diff
@@ -0,0 +1,192 @@
+        /// <param name="parameters">The parameters.</param>
+        /// <param name="nameMap">Maps a given property name to another one defined in the map.</param>
+        /// <remarks>Does not support collection parameters currently. Does not cache reflection results.</remarks>
+        public static void ReflectionParameterProvider<T>(this IDbCommand command, T parameters, IReadOnlyDictionary<string, string> nameMap = null)
```
The following functions handle parameter or return value objects in queries by reflection and either read or fill in the values by reflection from the database by matching a name. Basically what other ORMs, such as Dapper, are doing.

veikkoeeva commented on src/Orleans/Storage/DictionaryExtensions.cs:
```diff
@@ -0,0 +1,69 @@
+    /// <summary>
+    /// Extensions methods to work with dictionaries.
+    /// </summary>
+    public static class DictionaryExtensions
```
Some utility functions I used. Perhaps these could be internal and/or placed elsewhere entirely.

veikkoeeva commented on src/Orleans/Storage/GenericRelationalStorage.cs:
```diff
@@ -0,0 +1,210 @@
+    /// A general purpose class to work with a given relational database.
+    /// </summary>    
+    [DebuggerDisplay("InvariantName = {InvariantName}, ConnectionString = {ConnectionString}")]
+    public class GenericRelationalStorage: IRelationalStorage
```
This is the core of vendor agnostic functionality. Only this is strictly needed. *Maybe* a more suitable name would ``AdoNetRelationalStorage``. Opinions accepted? @jthelin?

jthelin replied:
Generic does not really feel like the right word because of the overloading with concepts of Generic<T>.

Looking at the usage patterns, I think you could just drop the prefix completely and call it simply `RelationalStorage`.

veikkoeeva commented on src/Orleans/Storage/GenericRelationalStorage.cs:
```diff
@@ -0,0 +1,210 @@
+
+        protected GenericRelationalStorage(string invariantName, string connectionString)
+        {
+            factory = DbProviderFactories.GetFactory(invariantName);
```
Here goes the invariant name. ADO.NET ``DbProviderFactories`` loads the appropriate dlls with the given provider invariant name. To be able to do so, it needs to know of them somehow first, either through GAC or by other configuration elements (see eariler comments).

veikkoeeva commented on src/Orleans/Storage/GenericRelationalStorage.cs:
```diff
@@ -0,0 +1,210 @@
+            {
+                connection.ConnectionString = connectionString;
+                await connection.OpenAsync().ConfigureAwait(continueOnCapturedContext: false);
+                using(var command = connection.CreateCommand())
```
Here, and see before, I've just split the various stages into separate functions and provided "a hook" via a supplied ``Func`` for the user to give parameters to query that needs to started first in order to create the vendor agnostic command object its parameters. Everything is async and configured without flowing the context.

veikkoeeva commented on src/Orleans/Storage/IRelationalStorage.cs:
```diff
@@ -0,0 +1,65 @@
+    /// <summary>
+    /// A common interface for all relational databases.
+    /// </summary>
+    public interface IRelationalStorage
```
I think providing an interface is prudent here. Maybe there will be a need to create a non-ADO.NET implementation. Maybe one wants to mock away database calls and return objects from elsewhere (e.g. an in-memory list).

jthelin replied:
+1

veikkoeeva commented on src/Orleans/Storage/OrleansRelationalConstants.cs:
```diff
@@ -0,0 +1,227 @@
+    /// <summary>
+    /// Query templates Orleans uses for operational queries.
+    /// </summary>
+    public class OrleansRelationalConstants
```
This is internal to Orleans. These are the keys that are used to distinquish the very finite amout of queries Orleans uses to interact with the database. I wonder why I have made this internal...

veikkoeeva commented on src/Orleans/Storage/OrleansRelationalExtensions.cs:
```diff
@@ -0,0 +1,1506 @@
+    /// Orleans specific, handcrafted convenience queries for efficiency.
+    /// </summary>
+    /// <remarks>This is public only to be usable to the statistics providers. Not intended for public use otherwise.</remarks>
+    public static class OrleansRelationalExtensions
```
Here are all the queries Orleans makes to the database. These would be simpler with reflection, e.g. there would be no need to create the parameter objects separately. Doing it like this spares some cycles from reflection and perhaps makes little less garbage.

veikkoeeva commented on src/Orleans/Storage/RelationalConstants.cs:
```diff
@@ -0,0 +1,379 @@
+    /// <summary>
+    /// A holder for well known, vendor specific connector class invariant names.
+    /// </summary>
+    public static class WellKnownRelationalInvariants
```
Here should be listed the invariant that are supported. Oracle is here as it supports a bit differing syntax on batch inserting multiple records with UNION ALL. Otherwise there isn't Oracle specific code as of yet. This is mostly used in testing, but all queries use this as the key (this also makes it possible to use multiple flavors of databases together provided one can synchronize them).

veikkoeeva commented on src/Orleans/Storage/RelationalConstants.cs:
```diff
@@ -0,0 +1,379 @@
+        /// The character that indicates a start escape key for columns and tables that are reserved words.
+        /// </summary>
+        /// <remarks>Vendor specific.</remarks>
+        public const string StartEscapeIndicatorKey = "StartEscapeIndicator";
```
For SQL Server these are square brackets, for MySQL these are backticks and so forth for other databases. These are used if/when queries need to be ran on tables that have reserved words are column names. More to do with general ORM specific functionality (though Orleans uses some reserved words too).

veikkoeeva commented on src/Orleans/Storage/RelationalConstants.cs:
```diff
@@ -0,0 +1,379 @@
+        /// A key for a query template to create a database with a given name.
+        /// </summary>
+        /// <remarks>Vendor specific.</remarks>
+        public const string CreateDatabaseKey = "CreateDatabaseTemplate";
```
These are used to make adding databases to tests more easily. Say, one wants to add a MySQL backend and test the membership implementation with it. Adding the needed queries to create and delete a database (used by tests) here makes it easy (should make, currently the tests are hardcoded to use SQL Server as there's no way to check provider invariant string from configs) as changing a connection string.

veikkoeeva commented on src/Orleans/Storage/RelationalConstants.cs:
```diff
@@ -0,0 +1,379 @@
+        /// A parameter indicator that can be used in queries which is then substituted according to the
+        /// underlying database.
+        /// </summary>
+        public const string InvariantParameterIndicator = "@";
```
When using reflection, the parameters need to be indicated somehow. Unfortunately the various databases have mostly their own indicators. Queries will work when using database specific indicators, but when using "invariant indicator", the queries should become vendor agnostic.

veikkoeeva commented on src/Orleans/Storage/RelationalConstants.cs:
```diff
@@ -0,0 +1,379 @@
+                  DELETE OrleansMembershipVersionTable;";
+
+            //Microsoft SQL Server.
+            AddOrModifyFactoryConstant(WellKnownRelationalInvariants.SqlServer, StartEscapeIndicatorKey, "[");
```
Here one can see SQL Server specific queries to create or drop a new database.

veikkoeeva commented on src/Orleans/Storage/RelationalManagementExtensions.cs:
```diff
@@ -0,0 +1,145 @@
+            //The timeout is regardless of what has been set on the command object itself and
+            //the query would be rolled back in the end. These defaults are more usable and
+            //can be customized per database.            
+            var transactionOptions = new TransactionOptions { IsolationLevel = IsolationLevel.ReadCommitted, Timeout = TransactionManager.MaximumTimeout };
```
With regard to async-await this is only useful starting from .NET 4.5.1 which allows for flowing the transaction context between threads (as tasks are likely running in different threads during their lifetime).

veikkoeeva commented on src/Orleans/Storage/RelationalStorageExtensions.cs:
```diff
@@ -0,0 +1,204 @@
+            //SqlParameters map is needed in case the query needs to be parameterized in order to avoid two
+            //reflection passes as first a query needs to be constructed and after that when a database
+            //command object has been created, parameters need to be provided to them.
+            var sqlParameters = new Dictionary<string, object>();
```
It might be smart to separate batch insert query creation from using it. Something to refactor later if this goes into Orleans.

veikkoeeva commented on src/Orleans/Storage/RelationalStorageUtilities.cs:
```diff
@@ -0,0 +1,69 @@
+        /// </summary>
+        /// <param name="sqlScript">The script from which to remove the separators.</param>
+        /// <returns>Scripts without separators.</returns>
+        public static IEnumerable<string> RemoveBatchSeparators(string sqlScript)
```
At the moment used by tests. They read the Orleans database script, batch it and (re-)create the database as defined in tests.

veikkoeeva commented on src/Orleans/Storage/SqlFormatProvider.cs:
```diff
@@ -0,0 +1,83 @@
+        }
+
+
+        private class SqlFormatter: ICustomFormatter
```
A formatter to format parameter to SQL queries in a well-defined matter in one place. This is needed especially in cases whe ``DbParameter`` objects are not used (Orleans uses currently parameters everywhere). One place to use these currently would be in tracing the queries.

veikkoeeva commented on src/Orleans/Storage/GenericRelationalStorage.cs:
```diff
@@ -0,0 +1,210 @@
+                    }
+
+                    command.CommandText = query;
+                    return await executor(command, selector).ConfigureAwait(continueOnCapturedContext: false);
```
It might be good to make ``GenericRelationalStorage`` to accept an Orleans logger and print both the query and paramers with their values here in some verbosity level.

veikkoeeva commented on src/OrleansProviders/SQLServer/CreateTables.sql:
```diff
@@ -1,113 +1,1440 @@
-DROP TABLE [dbo].[OrleansStatisticsTable]
-
-CREATE TABLE [dbo].[OrleansMembershipTable]
+﻿/*
```
This, along with ``GenericRelationalStorage`` is the gist of the matter. Most of the design is already commented. A point worth making is that Orleans has a dependency only to the names of the queries and the parameter names and types the queries use. That means, for instance, one can change table name or change functionality so as to, for example, remove ``UPDATE`` clauses as such and insert a new row provided the queries won't see this.

I have a plan to use this also to do queuing etc. Doing it like this doesn't make Orleans dependent on any one vendor and allows for vendor and deployment specific tuning. Also, the only used queries are presented here, there is no need to dig into code.

veikkoeeva commented on src/OrleansProviders/SQLServer/CreateTables.sql:
```diff
@@ -1,113 +1,1440 @@
+        WHEN LEFT(CAST(SERVERPROPERTY('ProductVersion') AS NVARCHAR), CHARINDEX('.', CAST(SERVERPROPERTY('ProductVersion') AS NVARCHAR)) - 1) = 10 
+            AND RIGHT(LEFT(CAST(SERVERPROPERTY('ProductVersion') AS NVARCHAR), 5), 2) = 50 THEN 'SQL Server 2008 R2' 
+        WHEN LEFT(CAST(SERVERPROPERTY('ProductVersion') AS NVARCHAR), CHARINDEX('.', CAST(SERVERPROPERTY('ProductVersion') AS NVARCHAR)) - 1) = 11 THEN 'SQL Server 2012'
+        WHEN LEFT(CAST(SERVERPROPERTY('ProductVersion') AS NVARCHAR), CHARINDEX('.', CAST(SERVERPROPERTY('ProductVersion') AS NVARCHAR)) - 1) = 12 THEN 'SQL Server 2014' 				
```
SQL Server 2016 is missing. :) I haven't checked if SQL Server Azure has something specific in it.

veikkoeeva commented on src/OrleansProviders/SQLServer/CreateTables.sql:
```diff
@@ -1,113 +1,1440 @@
+	(
+		'InsertMembershipKey',
+		N'SET NOCOUNT ON;
+		BEGIN TRANSACTION; --  @@TRANCOUNT = 0 -> +1.
```
This way of doing transaction handling is explicit and SQL Server 2000 compatible. It might be more clear to use try-catch in SQL Server 2005 and newer, maybe not. At least this is explicit on what is happening.

veikkoeeva commented on src/OrleansProviders/SQLServer/CreateTables.sql:
```diff
@@ -1,113 +1,1440 @@
+		IF @@TRANCOUNT > 0
+		BEGIN
+			COMMIT TRANSACTION;
+			SELECT CAST(1 AS BIT);
```
As stated eariler, I chose to be specific about this. One might get the same information also via "rows affected" or other means, but it would be more fragile as any other measure would rely on observing the success indirectly.

Notable here is that rather than returning just ``TRUE`` or ``FALSE`` anything else could be provided too. For instance, a reason for failure, like ``etag`` mismatch-

veikkoeeva commented on src/OrleansProviders/SQLServer/CreateTables.sql:
```diff
@@ -1,113 +1,1440 @@
+				[Timestamp]	= SYSUTCDATETIME(),
+				[Version]	= @version
+			WHERE
+				([DeploymentId]	= @deploymentId AND @deploymentId IS NOT NULL)
```
Having ``NULL`` valued parameters should never happen. Because the behaviour of ``NULL`` in SQL is difficult to predict (in practice one can treat as ``undefined`` in three-valued logic), I chose to be explicit about checking this case. It also documents the queries better, I feel.

Also, one could also test database functionality by providing the queries with parameters and see the results are what is expected and in that case querying would be less hazardous.

veikkoeeva commented on src/OrleansProviders/SQLServer/SqlStatisticsPublisher.cs:
```diff
@@ -21,20 +21,20 @@ MIT License
 namespace Orleans.Providers.SqlServer
 {
-    public class SqlStatisticsPublisher : IConfigurableStatisticsPublisher, IConfigurableSiloMetricsDataPublisher, IConfigurableClientMetricsDataPublisher, IProvider
+    public class SqlStatisticsPublisher: IConfigurableStatisticsPublisher, IConfigurableSiloMetricsDataPublisher, IConfigurableClientMetricsDataPublisher, IProvider
```
This class felt a bit messy with all the ``Init`` functionality and so forth. I'm not sure if there's a better way to do logging too. Ideas gladly welcomed here.

veikkoeeva commented on src/OrleansProviders/SQLServer/SqlStatisticsPublisher.cs:
```diff
@@ -74,210 +82,122 @@ public void AddConfiguration(string deployment, bool silo, string siloId, SiloAd
-
-            using (var conn = new SqlConnection(bldr.ConnectionString))
+            var siloOrClientName = (isSilo) ? siloName : clientId;
+            var id = (isSilo) ? siloAddress.ToLongString() : string.Format("{0}:{1}", siloOrClientName, generation);            
```
Maybe these have already a common function somewhere? Maybe formatting these should be refactored to functions of their own and given a descriptive name. It feels like ``id`` isn't a descriptive name for the variable either, but that's how it's all the way to the database currently (in general these variable names in code could be refactored to match exactly the ones in DB, currently they mostly are, but not all).

veikkoeeva commented on src/OrleansProviders/SQLServer/CreateTables.sql:
```diff
@@ -1,113 +1,1440 @@
-
-CREATE TABLE [dbo].[OrleansMembershipTable]
+﻿/*
+Project Orleans Cloud Service SDK ver. 1.0
```
I think the name of the script should be changed too. Perhaps something like ``CreateOrleansTables_SqlServer.sql``?

jthelin commented on src/Orleans/Storage/OrleansRelationalExtensions.cs:
```diff
@@ -0,0 +1,1506 @@
+
+            //All the rows have the same Version (.Item3) and VersionETag (.Item4) information.
+            //If .Item1 of the first element has no data, the MembershipEntry collection is empty.
+            var membershipData = ret.First().Item1 == null ? Enumerable.Empty<Tuple<MembershipEntry, string>>() : ret.Select(i => Tuple.Create(i.Item1, i.Item2));
```
Multiple enumerations of IEnumerable. Introduce variable for ret.First().

+ same pattern is several other places in this file.

[Found by ReSharper]

jthelin commented on src/Orleans/Storage/RelationalStorageExtensions.cs:
```diff
@@ -0,0 +1,204 @@
+            const string insertIntoValuesTemplate = "INSERT INTO {0} ({1}) SELECT {2};";
+            string columns = string.Empty;
+            var values = new List<string>();
+            if(parameters.Any())
```
Multiple enumerations of IEnumerable. Introduce variable for parameters.ToList() [or equivalent]

[Found by ReSharper]

jthelin commented on src/Orleans/Storage/OrleansRelationalExtensions.cs:
```diff
@@ -0,0 +1,1506 @@
+            //to BEGIN TRANSACTINION; INSERT INTO [OrleansStatisticsTable] <columns> SELECT <variables>; UNION ALL <variables> COMMIT TRANSACTION;
+            //where the UNION ALL is multiplied as many times as there are counters to insert.
+            int startFrom = queryTemplate.IndexOf("SELECT") + "SELECT".Length + 1; //This +1 is to have a space between SELECT and the first parameter name to not to have a SQL syntax error.
+            int lastSemicolon = queryTemplate.LastIndexOf(";");
```
IndexOf and LastIndexOf are cultural specific string comparisons. 
I think you want to make them use StringComparison.Ordinal.

[Found by ReSharper scan]

jthelin commented on src/Orleans/Storage/RelationalStorageExtensions.cs:
```diff
@@ -0,0 +1,204 @@
+                    {
+                        if(useSqlParams)
+                        {
+                            var parameterName = string.Format(string.Format(indexedParameterTemplate, parameterCount));
```
Duplicate string.Format call.

[Found by ReSharper scan]

jthelin commented on src/Orleans/Storage/OrleansRelationalExtensions.cs:
```diff
@@ -0,0 +1,1506 @@
+                string roleName = record.GetValue<string>("RoleName");
+                string instanceName = record.GetValue<string>("InstanceName");
+                int? updateZone = record.GetValue<int?>("UpdateZone");
+                int? faultZone = record.GetValueOrDefault<int?>("FaultZone");
```
ReSharper flags this as unused variable, but is really missing assignment into MembershipEntry.FaultZone at line 832 below.

jthelin commented on src/OrleansProviders/OrleansProviders.csproj:
```diff
@@ -100,7 +100,7 @@
   <ItemGroup>
     <Content Include="SQLServer\CreateTables.sql">
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
```
Any particular reason why this needs to be Copy Always?

Using PreserveNewest rather than Always gives much faster incremental build times overall.

jthelin commented:
Overall, looks great @veikkoeeva , modulo a few minor comments above.

One meta-question though: 

Given all the recent assembly restructuring to remove Azure / ZooKeeper dependencies from the core runtime assemblies, can and should we do something similar with the ADO.NET / SQL dependencies? 

This is really more related to the dependency levels on CoreCLR vs. CoreFx, so is not something we necessarily need to address with this PR, but I suspect as we look deeper into CoreCLR compatibility will be something we need to revisit.

However, I am NOT advocating delaying merge of this PR for "restructuring" unless it is very easy task. Better to do separate follow-on PR to deal with that separately, if others think it is important short-term objective.

veikkoeeva commented:
@jthelin Thanks for looking into this and for the comments. You have a point in all of them, I'll fix these in about 13-14 hours (kids etc.). I won't squash the commits until someone gives a go the master, so the changes are easier to spot.

Maybe we should also pick #557 on board before merging as I have made a few changes to how SQL Server tests are run and this might be the least overall work (less finding out for what to do for other people). What you and @gabikliot think?

For your meta-question, I'm all for separating this functionality to a separate dll. It feels like the way to go. I would think it's easier to merge this and then move it in a separate PR when someone has a place in mind, but I'm open to all ideas. I think I'll ask about small improvements and fixes in separate PRs otherwise too.

jthelin commented:
Yes, I think merging #557 before this one would be good extra safety net :)

sergeybykov commented:
I just merged https://github.com/dotnet/orleans/pull/557.

veikkoeeva commented:
Made the fixes and minor other modifications (e.g. renaming the SQL script). If this looks like being all right, I can  squash the commits. I'm off to sleep soon, but I'm here again in about 7-8 hours.

jthelin commented:
Looks good. Just one additional minor comment above, but if you want to fix that and squash then I think we can merge to master.

Nice job @veikkoeeva !

veikkoeeva commented:
@jthelin Heh, that was a good catch. These reviews sure are a life-saver. Fixed that, tests pass and commits squashed. If nobody finds anything, I'm OK with this and write some issues with a bit better time (probably fix some too). My thanks. :+1:

jasonwilliams200OK commented:
:+1: 
There were some extra linefeeds added in the code which can be cleaned up and some files are missing linefeed at the end; to avoid wrong invalid end-of-file terminator. (minor cosmetics, but it keeps the repository structure consistent) :smile:

veikkoeeva commented:
@jasonwilliams200OK Hmm, I do actually use two linefeeds on some places (end of usings, between functions). :) Likewise always curly braces and put parenthesis next to conditionals etc. It doesn't look lik there is one consistent style in the repo, maybe there should, but maybe it should be fixed in one gigantic refactoring (with R# or somesuch perhaps?).

jthelin commented:
I'll merge this change, and then Gabi has a few minor code cleanup tweaks for code consistency which he will do as separate PR, much like he did with the recent new ZK feature. 
We'll sort our @jasonwilliams200OK feedback comments in that PR.

Thanks for the contribution! :+1:

sergeybykov commented:
@veikkoeeva This was a major multi-week effort. Thank you! I hope no children were hurt in the process. :-)

gabikliot commented:
Thanks you very much @veikkoeeva ! That has been quite an BIG effort! Our SQL story is much more solid now thanks to you.

veikkoeeva commented:
You are welcome. No children were hurt (which made it a multi-week effort, but not too many hours). :)

pherbel commented:
Congrats @veikkoeeva!  Nice contribution!
